pkg/services/secrets/manager/manager_test.go

```diff
@@ -401,3 +408,100 @@ func TestSecretsService_Decrypt(t *testing.T) {
 		assert.Equal(t, []byte("grafana"), decrypted)
 	})
 }
+
+func TestIntegration_SecretsService(t *testing.T) {
+	if testing.Short() {
+		t.Skip("skipping integration test")
+	}
+
+	ctx := context.Background()
+	someData := []byte(`some-data`)
+
+	tcs := map[string]func(*testing.T, *sqlstore.SQLStore, *SecretsService){
+		"regular": func(t *testing.T, _ *sqlstore.SQLStore, svc *SecretsService) {
+			_, err := svc.Encrypt(ctx, someData, secrets.WithoutScope())
+			require.NoError(t, err)
+		},
+		"within successful InTransaction": func(t *testing.T, store *sqlstore.SQLStore, svc *SecretsService) {
+			require.NoError(t, store.InTransaction(ctx, func(ctx context.Context) error {
+				_, err := svc.Encrypt(ctx, someData, secrets.WithoutScope())
+				require.NoError(t, err)
+				return nil
+			}))
+		},
+		"within unsuccessful InTransaction": func(t *testing.T, store *sqlstore.SQLStore, svc *SecretsService) {
+			require.NotNil(t, store.InTransaction(ctx, func(ctx context.Context) error {
+				_, err := svc.Encrypt(ctx, someData, secrets.WithoutScope())
+				require.NoError(t, err)
+				return errors.New("error")
+			}))
+		},
+		"within unsuccessful InTransaction (plus forced db fetch)": func(t *testing.T, store *sqlstore.SQLStore, svc *SecretsService) {
+			require.NotNil(t, store.InTransaction(ctx, func(ctx context.Context) error {
+				encrypted, err := svc.Encrypt(ctx, someData, secrets.WithoutScope())
+				require.NoError(t, err)
+				svc.dataKeyCache.flush()
+				decrypted, err := svc.Decrypt(ctx, encrypted)
+				require.NoError(t, err)
+				assert.Equal(t, someData, decrypted)
+				return errors.New("error")
+			}))
+		},
```

Comment:
This test case reproduces what's being discussed [here](https://github.com/grafana/grafana-enterprise/issues/4252#issuecomment-1359672826) (cc/ @leandro-deveikis).

That's why I also need to modify the cache implementation.